sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -201,11 +222,30 @@ public Read withMaxReadTime(Duration maxReadTime) {
       return toBuilder().setMaxReadTime(maxReadTime).build();
     }
 
+    /**
+     * Specifies how late records consumed by this source can be to still be considered on time.
+     * When this limit is exceeded the actual backlog size will be evaluated and the runner might
+     * decide to scale the amount of resources allocated to the pipeline in order to
+     * speed up ingestion.
+     *
+     * <p>
+     * TODO: This feature will not work properly with current {@link KinesisReader#getWatermark()}
```

Comment:
Not sure I understand this comment. It seems to say that the feature is currently working incorrectly - in that case, I'd suggest to remove the feature and re-add it when it's possible to make it work correctly, because there's no benefit to the user in having the API expose a feature that doesn't work. However, you say above that you've been running this and it was working for you. Can you clarify?

Reply:
Thanks @jkff for your review. This feature works for us in our custom Kinesis connector implementation for Dataflow SDK where we have implemented some features required for production usage. Now I'm trying to port these features to Beam SDK ([BEAM-2455](https://issues.apache.org/jira/browse/BEAM-2455), [BEAM-2467](https://issues.apache.org/jira/browse/BEAM-2467), [BEAM-2468](https://issues.apache.org/jira/browse/BEAM-2468), [BEAM-2469](https://issues.apache.org/jira/browse/BEAM-2469)). We already have all the required code but it just seemed a better option to split the changes into few pull requests for clarity. Now I see that I should submit BEAM-2467 before this one but as I explained to Raghu, I noticed the dependency after submitting this PR.

So if it's not a problem I would leave this feature with TODO comment that will be fixed in BEAM-2467.

Reply:
Thanks. If the current PR does not work at all until BEAM-2467 is implemented, then let us revisit this PR after it is implemented: I don't see value to a user in committing this PR before then. Is there currently a PR in review for BEAM-2467? I couldn't find one.

Reply:
@jkff I just submitted a PR for BEAM-2467 here: https://github.com/apache/beam/pull/3851
So let's try to close that one quickly and I'll get back to this